KG-440. Handle Assistant message in subgraphWithTask method

sdubov opened the pull request:
[KG-440](https://youtrack.jetbrains.com/issue/KG-440/Update-the-subgraphWithTask-to-use-giveFeedback-node-with-assistant-response). SubgraphWithTask API supports models with no "tool_choice" capability.

- Add ability to use `subgraphWithTask` API with models that does not have `tool_choice` capability;
- Update mock test framework logic to be able to set conditions on tool call mocks.

## Breaking Changes
No

---

#### Type of the changes
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Tests improvement
- [ ] Refactoring

#### Checklist
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [x] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [x] An issue describing the proposed change exists
- [x] The pull request includes a link to the issue
- [x] The change was discussed and approved in the issue
- [ ] Docs have been added / updated

sdubov committed:
KG-440. Update subgraphWithTask API to handle assistant message

- For models that do not have tool_choice capability, use updated prompt to force LLM to use tool calls instead of Assistant response. For models with tool_choice, fail on Assistant message.

sdubov committed:
KG-440. Add ability to set condition for tool calls mock in mock LLM executor

antoniibelyshev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -310,33 +316,37 @@ public inline fun <reified Input, reified Output, reified OutputTransformed> AIA
     noinline defineTask: suspend AIAgentGraphContextBase.(Input) -> String
 ) {
+    val askAssistantToFinishMaximum = SubgraphWithTaskUtils.ASSISTANT_RESPONSE_REPEAT_MAX
+    var askAssistantToFinishCounter = 1
```
It's better to put it in the storage during the setup and update there. Otherwise it could cause concurrency issues: for example, when using parallel node execution.

sdubov replied:
@antoniibelyshev, could you please explain your concern here a bit? When running nodes in parallel, I would like the code to be isolated in each thread. E.g. when I run two `subgraphWithTasks` in parallel, I would like them to handle assistant messages independently as I can run them on different models, for example.

antoniibelyshev replied:
As I understand it, one could create an instance of the `subgraphWithTask` node, and then pass it twice to the parallel node:

```
val task by nodeDoNothing<Unit>()
val parallelNode by parallel(task, task) { ... }
```

then both branches will access the same instance of `askAssistantToFinishCounter`.

As far as I remember, this concern was initially raised by @EugeneTheDev when I worked on the retry subgraph. Probably, he could help here.

EugeneTheDev replied:
Yes, this is an easy to make mistake, you can't use variables inside your strategy just like this. The reason is that strategy instance can be shared between different agents, since this is a sort of a blueprint, and this would make this counter state shared too. If you want to store some state during strategy execution, you have to use `storage` map, since it is tied to the execution context, not to the strategy itself. 

But it won't protect from concurrency issues, since this is not the concern of this `storage` component. It sounds like you just need to avoid passing the same node several times if you want to do a parallel run, just create two copies.

sdubov replied:
Fixed. Please see the last commit with the fix.

antoniibelyshev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -364,24 +376,60 @@ public inline fun <reified Input, reified Output, reified OutputTransformed> AIA
+    @OptIn(DetachedPromptExecutorAPI::class)
+    val handleAssistantMessage by node<Message.Assistant, Message.Response> { response ->
+        if (llm.model.capabilities.contains(LLMCapability.ToolChoice)) {
+            error(
```
Let's not not throw an error, and just give a warning in the logs. I would say, there could be a scenario when a model with `LLMCapability.ToolChoice` does not generate a tool. We should give it a chance to recover.

sdubov replied:
Thank you for pointing this out. It was the original logic. My expectation here was: if a model has `tool_choice` capability and we ask it to call tools, there is no way it return anything else, but a tool call. That's why we throw here. 

I've discussed this originally with @tiginamaria. @antoniibelyshev, could you please provide some examples or confirm that a model can return an Assistant message back instead?

antoniibelyshev replied:
Generally, I also would expect the model with tool_choice to always produce tools when asked. However, I am not totally sure it is the case for all providers. Anyway, if we got to this point, the response was not a tool call, meaning that the model does not produce only tool calls

EugeneTheDev replied:
Can we handle the case when the model doesn't have tool calling capability early on, doing this validation as a first step in this function? If we know that model would never produce tool calls, we can avoid spending tokens on requesting it to provide assistant message, if we are going to throw exception later on anyways, without using this assistant message in any meaningful way.

antoniibelyshev replied:
If the model does not have ToolChoice capability, it still might call tools, so it makes sense to correct it only when it didn't call a tool.

sdubov replied:
@EugeneTheDev, that was the whole point for this change. `subgraphWithTask` API should support any model that a user would like to use.

antoniibelyshev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -364,24 +376,60 @@ public inline fun <reified Input, reified Output, reified OutputTransformed> AIA
+            updatePrompt {
+                user {
+                    markdown {
+                        h1("DO NOT CHAT WITH ME DIRECTLY! CALL TOOLS, INSTEAD.")
```
Can we allow to customize the feedback, make it a parameter with default value? The models which are weak at tool calling require more accurate prompting, which could also be model-specific, so we can let the users tweak this by themselves. WDYT?

sdubov replied:
@antoniibelyshev, what other feedback prompts do you have in mind? Maybe we can generalise it to keep it up to the library? If it is not the case, I can add a parameter as well.

antoniibelyshev replied:
I do not have specific prompts in mind

EugeneTheDev replied:
+1 here and in general for making all prompts customizable with default value (you can check e.g. how it is done for `JsonStructuredData`)

sdubov replied:
Discussed this in person. Agreed with @EugeneTheDev  to keep it internal and update it later if we need to customise it from any perspective.

antoniibelyshev reviewed (approved):
Thank you, looks good! Just pointed out a few minor issues

EugeneTheDev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -310,33 +316,37 @@ public inline fun <reified Input, reified Output, reified OutputTransformed> AIA
     finishToolDescriptor: ToolDescriptor,
     noinline defineTask: suspend AIAgentGraphContextBase.(Input) -> String
 ) {
+    val askAssistantToFinishMaximum = SubgraphWithTaskUtils.ASSISTANT_RESPONSE_REPEAT_MAX
```
Can this be moved to function parameters with default value, to make it configurable?

sdubov replied:
@EugeneTheDev, sure we can move it to parameters. However, my point here was that it is the internal implementation that I would explicitly want to hide from a user. From my perspective, it is not something that a user should configure. It is a parameter that we need to define internally to make sure our `subgraphWithTask` works well.

sdubov replied:
Fixed.

EugeneTheDev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -364,24 +376,60 @@ public inline fun <reified Input, reified Output, reified OutputTransformed> AIA
     nodeStart then setupTask then nodeCallLLM then nodeDecide
 
     edge(nodeDecide forwardTo callToolHacked onToolCall { true })
+    edge(nodeDecide forwardTo handleAssistantMessage onCondition { it is Message.Assistant } transformed { it as Message.Assistant })
```
nit: You can combine condition and transformation using `onIsInstance`

sdubov replied:
Fixed. Thanks for there nit!

sdubov committed:
KG-440. Fix the serialization issue in subgraphWithTask

- Update the subgraph logic to execute a "real" finish tool to get a result instead of get a result by casting the finish tool arguments.

sdubov commented:
@EugeneTheDev, fixed all comments based on our coding session today. Could you please check the last commit in the PR?

EugeneTheDev commented on agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/agent/AIAgentSubgraphExt.kt:
```diff
@@ -301,42 +341,46 @@
 public inline fun <reified Input, reified Output, reified OutputTransformed> AIAgentSubgraphBuilderBase<Input, OutputTransformed>.setupSubgraphWithTask(
-    finishToolDescriptor: ToolDescriptor,
+    finishTool: Tool<Output, OutputTransformed>,
+    assistantResponseRepeatMax: Int? = null,
```
nit: Why do we make it nullable and use constants as a fallback (ASSISTANT_RESPONSE_REPEAT_MAX). Why can we just get rid of the constant and nullable parameter and make it just a default value - `assistantResponseRepeatMax: Int = 3`? Should look more idiomatic

sdubov replied:
I prefer to have one point when I resolve the default parameters. If you have multiple invocations where you need to pass an optional parameters, then you need to specify the default parameter in every method:
```
fun a(foo: Int = 3) = b(foo)
fun b(foo: Int = 3) { ... }
```
I think it is easier and less error prone when you have a logic that resolve the default parameters in one place (root function).

EugeneTheDev replied:
Good point, then you can still have a const that you just assign as a default parameter everywhere - no need to have nullable parameters if the actual value is always not null (number of iterations)

EugeneTheDev replied:
AFAIK this is a common practice, e.g.`kotlinx.coroutines` do that too. You can take a look at how e.g. `Channel` (or other similar functions) are defined

EugeneTheDev reviewed (approved):
LGTM, thanks, minor comment only

sdubov committed:
Move messages about nodes execution inside subgraph under debug logs level
